pkg/common/templates/generic_lookup_func.go

```diff
@@ -0,0 +1,161 @@
+// Copyright (c) 2020 Red Hat, Inc.
+// Copyright Contributors to the Open Cluster Management project
+
+package templates
+
+import (
+  metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+  "github.com/golang/glog"
+  apierrors "k8s.io/apimachinery/pkg/api/errors"
+  "k8s.io/client-go/dynamic"
+  "k8s.io/client-go/discovery"
+  "k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+
+
+func lookup (apiversion string, kind string, namespace string, rsrcname string) (map[string]interface{}, error) {
+  glog.V(2).Infof("lookup :  %v, %v, %v, %v", apiversion, kind, namespace, rsrcname)
+
+  result  := map[string]interface{}{}
+
+  //get dynamic Client for the given GVK and namespace
+	dclient,  dclientErr := getDynamicClient(apiversion, kind, namespace)
+  if(dclientErr != nil){
+    return result, dclientErr
+  }
+
+  //if resourcename is  set then get the specific resource
+  //else get list of all resources for that (gvk, ns)
+
+  var lookupErr error
+	if rsrcname != "" {
+    getObj, getErr := dclient.Get( rsrcname, metav1.GetOptions{})
+    if getErr == nil {
+      result = getObj.UnstructuredContent()
+    }
+    lookupErr = getErr
+  }else{
+    listObj, listErr := dclient.List( metav1.ListOptions{})
+    if listErr == nil {
+      result = listObj.UnstructuredContent()
+    }
+    lookupErr = listErr
+  }
+
+  if lookupErr != nil {
+    if apierrors.IsNotFound(lookupErr) {
+        lookupErr = nil
+    }
+  }
+
+  glog.V(2).Infof("lookup result:  %v", result)
+  return result, lookupErr
+}
+
+
+//this func finds the GVR for given GVK and returns a namespaced dynamic client
+func getDynamicClient(apiversion string, kind string, namespace string) (dynamic.ResourceInterface,  error) {
+
+  var dclient dynamic.ResourceInterface
+	gvk := schema.FromAPIVersionAndKind(apiversion, kind)
+  glog.V(2).Infof("GVK is:  %v", gvk)
+
+  // we have GVK but We need GVR i.e resourcename for kind inorder to create dynamicClient
+  // find ApiResource for given GVK
+	apiResource, findErr := findAPIResource(gvk)
+	if findErr != nil {
+		return nil,  findErr
+	}
+  //make GVR from ApiResource
+	gvr := schema.GroupVersionResource{
+		Group:    apiResource.Group,
+		Version:  apiResource.Version,
+		Resource: apiResource.Name,
+	}
+  glog.V(2).Infof("GVR is:  %v", gvr)
+
+  //get Dynamic Client
+	dclientIntf , dclientErr := dynamic.NewForConfig(kubeConfig)
+	if dclientErr != nil {
+    glog.Errorf("Failed to get dynamic client with err: %v", dclientErr)
+		return nil,  dclientErr
+	}
+
+  //get Dynamic Client for GVR
+	dclientNsRes := dclientIntf.Resource(gvr)
+
+  //get Dynamic Client for GVR for Namespace if namespaced
+  if apiResource.Namespaced && namespace != "" {
+    dclient = dclientNsRes.Namespace(namespace)
+  }else{
+    dclient = dclientNsRes
+  }
+
+  glog.V(2).Infof("dynamic client:  %v", dclient)
```

Comment:
should this be cleaned up? looks like logs for debugging

Reply:
by default Info is not logged and it is also at level 2. .V(2).Infof . So this would appear in the logs only if log level is elevated to level 2 and Info is enabled